File: Civitas ruby/lib/sorpresa.rb
module Civitas
class Sorpresa
  
  def initialize(tipo, tablero, valor, texto, mazo)
    @tipo = tipo
    @tablero = tablero
    @valor = valor
    @texto = texto
    @mazo = mazo
  end
  
  def self.tableroSolo(tipo, tablero)
    self.new(tipo, tablero, -1, nil, nil)
  end
  
  def self.tableroTexto(tipo, tablero, valor, texto)
    self.new(tipo, tablero, valor, texto, nil)
  end
  
  def self.valor(tipo, valor, texto)
    self.new(tipo, nil, valor, texto, nil)
  end
  
  def self.mazo(tipo, mazo)
    self.new(tipo, nil, -1, nil, mazo)
  end
  
  
  
  
  private
  
   def aplicaAJugador_irACasilla(actual, todos)
    if(jugadorCorrecto(actual, todos))
      jugador = todos[actual]
      informe(actual, todos)
      casillaActual = jugador.getNumCasillaActual
      tirada = @tablero.calcularTirada(casillaActual, @valor)
      nuevaPosicion = @tablero.nuevaPosicion(casillaActual, tirada)
      jugador.moverACasilla(nuevaPosicion)
      @tablero.getCasilla(nuevaPosicion).recibeJugador(actual, todos)
    end
  end
  
  def aplicarAJugador_irACarcel(actual, todos)
    if(jugadorCorrecto(actual, todos))
      jugador = todos[actual]
      informe(actual, todos)
      juador.encarcelar(tablero.getCarcel())
    end
  end
  
  def aplicarAJugador_pagarCobrar(actual, todos)
    if(jugadorCorrecto(actual, todos))
      informe(actual, todos)
      jugador = todos[actual]
      jugador.modificarSaldo(@valor)
    end
  end
  
  def aplicarAJugador_porCasaHotel(actual, todos)
    if(jugadorCorrecto(actual, todos))
      informe(actual, todos)
      jugador = todos[actual]
      numCasasHoteles = jugador
      jugador.modificarSaldo(valor*jugador.cantidadCasasHoteles)
    end
  end
  
  def aplicarAJugador_porJugador(actual, todos)
    if(jugadorCorrecto(actual, todos))
      informe(actual, todos)
      jugador = todos[actual]
      sorpresa1 = valor.new(Tipo_sorpresa::PAGARCOBRAR, @valor*-1, nil)
      
      for i in (1..todos.length)
        if(todos[i] != actual)
          sorpresa1.aplicarAJugador_pagarCobrar(i, todos)
        end
      end
      
      sorpresa2 = valor.new(Tipo_sorpresa::PAGARCOBRAR, @valor*(todos.length - 1), nil)
      sorpresa2.aplicarAJugador_pagarCobrar(actual, todos)
    end
  end
  
  def aplicarAJugador_salirCarcel(actual, todos)
    if(jugadorCorrecto(actual, todos))
      encuentra = false
      informe(actual, todos)
      
      for i in (1..todos.length)
        if(todos[i].tieneSalvoConducto)
          encuentra = true
        end
      end
      
      if(!encuentra)
        todos[actual].obtenerSalvoConducto
        salirDelMazo
      end
    end
  end
      
  def informe(actual, todos)
    Diario.instance.ocurre_evento("\nSe aplica sorpresa al jugador: " + todos[actual].getNombre)
  end
  
  
  
  
  
  public
  
  def jugadorCorrecto(actual, todos)
    if(actual < todos.length)
      return true
    else
      return false
    end
  end
  
  def toString
    return @texto
  end
  
  
    
  def aplicarAJugador(actual, todos)
    if(@tipo == Tipo_sorpresa::IRCASILLA)
      aplicarAJugador_irACasilla(actual, todos)
    elsif (@tipo == Tipo_sorpresa::IRCARCEL)
      aplicarAJugador_irACarcel(actual, todos)
    elsif (@tipo == Tipo_sorpresa::PAGARCOBRAR)
      aplicarAJugador_oagarCobrar(actual, todos)
    elsif (@tipo == Tipo_sorpresa::PORCASAHOTEL)
      aplicarAJugador_porCasaHotel(actual, todos)
    elsif (@tipo == Tipo_sorpresa::PORJUGADOR)
      aplicarAJugador_porJugador(actual, todos)
    elsif (@tipo == Tipo_sorpresa::SALIRCARCEL)
      aplicarAJugador_salirCarcel(actual, todos)
    else
      puts("\n\n\nTipo de sorpresa erronea")
    end
  end
  
  def salirDelMazo
    if(@tipo == Tipo_sorpresa::SALIRCARCEL)
      @mazo.inhabilitarCartaEspecial(self)
    end
  end
  
  def usada
    if(@tipo == Tipo_sorpresa::SALIRCARCEL)
      @mazo.habilitarCartaEspecial(self)
    end
  end
end
end

File: Civitas java/src/civitas/Sorpresa.java
package civitas;

import java.util.ArrayList;

public class Sorpresa{
    private String texto;
    private int valor;
    private Tablero tablero;
    private MazoSorpresas mazo;
    private TipoSorpresa tipo;    
        
    Sorpresa(TipoSorpresa tipo1, Tablero tablero1){
        init();
        tipo = tipo1;
        tablero = tablero1;
    }
    
    Sorpresa(TipoSorpresa tipo1, Tablero tablero1, int valor1, String texto1){
        init();
        tipo = tipo1;
        tablero = tablero1;
        valor = valor1;
        texto = texto1;
    }
    
    Sorpresa(TipoSorpresa tipo1, int valor1, String texto1){
        init();
        tipo = tipo1;
        valor = valor1;
        texto = texto1;
    }
    
    Sorpresa(TipoSorpresa tipo1, MazoSorpresas mazo1){
        init();
        tipo = tipo1;
        mazo = mazo1;
    }
    
    void aplicarAJugador(int actual, ArrayList<Jugador> todos){
        if(jugadorCorrecto(actual, todos)){
            if(tipo == TipoSorpresa.IRCARCEL)
                aplicarAJugador_irACarcel(actual, todos);
            else if(tipo == TipoSorpresa.IRCASILLA)
                aplicarAJugador_irACasilla(actual, todos);
            else if(tipo == TipoSorpresa.PAGARCOBRAR)
                aplicarAJugador_pagarCobrar(actual, todos);
            else if(tipo == TipoSorpresa.PORCASAHOTEL)
                aplicarAJugador_porCasaHotel(actual, todos);
            else if(tipo == TipoSorpresa.PORJUGADOR)
                aplicarAJugador_porJugador(actual, todos);
            else if(tipo == TipoSorpresa.SALIRCARCEL)
                aplicarAJugador_salirCarcel(actual, todos);
            else 
                System.out.println("AplicaraJugador ha dadao error");
        }
    }
    
    private void aplicarAJugador_irACasilla(int actual, ArrayList<Jugador> todos){
        informe(actual, todos);
        int casillaActual = todos.get(actual).getNumCasillaActual();
        tablero.calcularTirada(casillaActual, valor);
    }
    
    private void aplicarAJugador_irACarcel(int actual, ArrayList<Jugador> todos){
        todos.get(actual).encarcelar(tablero.getCarcel());
    }
    
    private void aplicarAJugador_pagarCobrar(int actual, ArrayList<Jugador> todos){
        informe(actual, todos);
        todos.get(actual).modificarSaldo(valor);
    }
    
    private void aplicarAJugador_porCasaHotel(int actual, ArrayList<Jugador> todos){
        informe(actual, todos);
        todos.get(actual).modificarSaldo(valor*todos.get(actual).cantidadCasasHoteles());    

    }
    
    private void aplicarAJugador_porJugador(int actual, ArrayList<Jugador> todos){
        informe(actual, todos);
        Sorpresa sorpresa1 = new Sorpresa(TipoSorpresa.PAGARCOBRAR, -1, "PAGARCOBRAR");
        
        for(int i = 0; i < todos.size(); i++){
            if(i != actual)
                sorpresa1.aplicarAJugador(i, todos);
        }
        
        Sorpresa sorpresa2 = new Sorpresa(TipoSorpresa.PAGARCOBRAR, valor*(todos.size() - 1), "PAGARCOBRAR");
        sorpresa2.aplicarAJugador(actual, todos);
    }
    
    private void aplicarAJugador_salirCarcel(int actual, ArrayList<Jugador> todos){
        informe(actual, todos);
        boolean tiene = false;
        
        for(int i = 0; i < todos.size(); i++){
            if(todos.get(actual).tieneSalvoconducto())
                tiene = true;
        }
        
        if(!tiene){
            todos.get(actual).obtenerSalvoConducto(this);
            salirDelMazo();
        }
    }
    
    private void informe(int actual, ArrayList<Jugador> todos){
        (Diario.getInstance()).ocurreEvento("Se aplica una sorpresa al jugador " + todos.get(actual).getNombre());
    }
    
    private void init(){
        valor = -1;
        mazo = null;
        tablero = null;
    }
    
    public boolean jugadorCorrecto(int actual, ArrayList<Jugador> todos){
        if(actual < todos.size())
            return true;
        else
            return false;
    }
    
    void salirDelMazo(){
        if(tipo == TipoSorpresa.SALIRCARCEL){
            System.out.println("TIPO SORPRESA SALIR DEL MAZO: " + tipo);
            mazo.inhabilitarCartaEspecial(this);
        }
    }
    
    public String toString(){
        return texto;
    }
    
    void usada(){
        if(tipo == TipoSorpresa.SALIRCARCEL)
            mazo.habilitarCartaEspecial(this);
    }
}
